test/Spec/BotPlutusInterface/Balance.hs

```diff
@@ -105,3 +165,131 @@ addUtxosForChange = do
+      case eRslt of
+        (Left txt) -> assertFailure ("PAB effect error: " <> Text.unpack txt)
+        (Right (Left txt)) -> assertFailure $ "Balancing error: " <> Text.unpack txt -- <> "\n(Tx: " <> show unbalancedTx <> ")"
+        (Right (Right trx)) -> do
+          let scrTxOut'' = scrTxOut' & Ledger.ciTxOutValue <>~ Ada.lovelaceValueOf 500
+              scrTxOutNew = Ledger.toTxOut scrTxOut''
+          assertBool
+            ( "Expected UTxO not in output Tx."
+                <> "\nExpected UTxO: "
+                <> show scrTxOutNew
+                <> "\nNew UTxOs: "
+                <> show (txOutputs trx)
+                <> "\nUnbalanced UTxOs: "
+                <> show (txOutputs (unbalancedTx ^. OffChain.tx))
+            )
+            (scrTxOutNew `elem` txOutputs trx)
+
+-- Like the first one, but
+-- only has inputs from the script.
+dontAddChangeToDatum2 :: Assertion
+dontAddChangeToDatum2 = do
+  let scrTxOut' =
+        ScriptChainIndexTxOut
+          valAddr
+          (Right validator) -- (valHash, Just validator)
+          (Right scrDatum) -- (scrDatumHash, Just scrDatum)
+          (scrValue <> Ada.lovelaceValueOf 1_500_000)
+      scrTxOut = Ledger.toTxOut scrTxOut'
+      -- initState :: MockContractState ()
+      initState =
+        def & utxos .~ [(txOutRef6, scrTxOut)]
+          & contractEnv .~ contractEnv'
+      pabConf :: PABConfig
+      pabConf = def {pcOwnPubKeyHash = pkh3}
+      -- contractEnv' :: ContractEnvironment ()
+      contractEnv' = def {cePABConfig = pabConf}
+
+      scrLkups =
+        Constraints.unspentOutputs (Map.fromList [(txOutRef6, scrTxOut')])
+          <> Constraints.ownPaymentPubKeyHash paymentPkh3
+      txConsts =
+        -- Pay the same datum to the script, but with LESS ada
+        -- and fewer tokens. This is to ensure that the excess
+        -- ADA and tokens are moved into their own UTxO(s),
+        -- (The extra ada is used to cover fees etc...)
+        Constraints.mustPayToOtherScript valHash scrDatum scrValue'
+          <> Constraints.mustSpendScriptOutput txOutRef6 Ledger.unitRedeemer
+      eunbalancedTx = Constraints.mkTx @Void scrLkups txConsts
+
+  case eunbalancedTx of
+    Left mkTxErr -> assertFailure ("MkTx Error: " <> show mkTxErr)
+    Right unbalancedTx -> do
+      let (eRslt, _finalState) = runPABEffectPure initState (balanceTxIO @() @'[PABEffect ()] pabConf pkh3 unbalancedTx)
+      case eRslt of
+        (Left txt) -> assertFailure ("PAB effect error: " <> Text.unpack txt)
+        (Right (Left txt)) -> assertFailure $ "Balancing error: " <> Text.unpack txt -- <> "\n(Tx: " <> show unbalancedTx <> ")"
+        (Right (Right trx)) -> do
+          let scrTxOut'' = scrTxOut' & Ledger.ciTxOutValue .~ scrValue'
+              scrTxOutNew = Ledger.toTxOut scrTxOut''
+          assertBool
+            ( "Expected UTxO not in output Tx."
```

Comment:
I'd suggest same change for failure reporting as the one proposed [here](https://github.com/mlabs-haskell/bot-plutus-interface/pull/140/files#diff-f259de0cbdc73d3a99698ed958a00bd0089f4ed1a068f151af4944b062a0ed96R234).

And also check that additional output was added to handle change.